### PR TITLE
extract files to dedicated properties to mark them with PathSensivity

### LINKED ISSUE #30: Plugin incompatible with Gradle 7.x due to missing input declarations
Gradle is complaining about missing input declarations as this will no longer be supported in Gradle 7 the plugin will become incompatible, These are the deprecation warnings gradle enterprise is printing out for a `com.github.eerohele.SaxonXsltTask`:

- Property 'advancedOptions' is not annotated with an input or output annotation.
- Property 'catalogManager' is not annotated with an input or output annotation.
- Property 'commonArguments' is not annotated with an input or output annotation.
- Property 'defaultOutputExtension' is not annotated with an input or output annotation.
- Property 'options' is not annotated with an input or output annotation.
- Property 'parameters' is not annotated with an input or output annotation.
- Property 'stylesheetParameters' is not annotated with an input or output annotation.

### PR DESCRIPTION
**Pull request by dgeissl**
- remove file based properties from options map and place them in dedicated variables (adding `@InputFile` `@PathSensitive` annotations)
-- input
-- catalog
-- stylesheet
-- output
- expose other properties as `@Input` properties
- cleanup redundant semicolons

this should fix: #30 and #29

### CODE REVIEW

File: src/main/groovy/com/github/eerohele/SaxonXsltTask.groovy [hunk 1]
```diff
@@ -26,22 +31,29 @@ class SaxonXsltTask extends DefaultTask {
     protected static final String PERIOD = '.'
     protected static final String XSLT_NAMESPACE = 'http://www.w3.org/1999/XSL/Transform'
 
-    protected final List<String> defaultArguments = ['-quit:off'].asImmutable()
+    protected static final List<String> DEFAULT_ARGUMENTS = ['-quit:off'].asImmutable()
 
-    protected final Map<String, Object> options = [output: project.buildDir]
+    private Object inputFiles
```
Comment (eerohele) [inline src/main/groovy/com/github/eerohele/SaxonXsltTask.groovy:36]: Is it necessary to move these things into variables for Gradle 7.1.1? I'd rather not do it if we don't have to. For one thing, this change is in conflict with the changes @ndw proposes in https://github.com/eerohele/saxon-gradle/pull/26.

File: src/main/groovy/com/github/eerohele/SaxonXsltTask.groovy [hunk 2]
```diff
@@ -175,459 +187,459 @@ class SaxonXsltTask extends DefaultTask {
     // START ADVANCED OPTIONS
 
     void allowExternalFunctions(Boolean x) {
-        this.advancedOptions['allow-external-functions'] = x;
```
Comment (eerohele) [inline src/main/groovy/com/github/eerohele/SaxonXsltTask.groovy:178]: I understand these semicolons are redundant, but if we want to remove it, we should definitely do it in a separate commit.

File: src/main/groovy/com/github/eerohele/SaxonXsltTask.groovy [hunk 3]
```diff
@@ -725,34 +737,64 @@ class SaxonXsltTask extends DefaultTask {
...
-    @OutputFiles
-    FileCollection getOutputFiles() {
-        if (this.options.input != null) {
-            project.files(project.files(this.options.input).collect { getOutputFile(it) })
-        } else {
-            project.files(this.options.output)
-        }
+    @InputFiles
+    @SkipWhenEmpty
+    @PathSensitive(PathSensitivity.RELATIVE)
```
Comment (eerohele) [inline src/main/groovy/com/github/eerohele/SaxonXsltTask.groovy:749]: I believe the `@PathSensitive` attribute is a fix to #29. That should be in a separate commit. Ideally, there'd be just one commit that fixes Gradle 7 compatibility issues and nothing else.

File: src/main/groovy/com/github/eerohele/SaxonXsltTask.groovy [hunk 4]
```diff
@@ -789,11 +831,11 @@ class SaxonXsltTask extends DefaultTask {
     // and output.
     @Internal
     protected List<String> getCommonArguments() {
-        Map<String, String> commonOptions = this.options.findAll { name, value ->
-            !['input', 'output'].contains(name)
-        }.asImmutable()
-
-        (commonOptions.inject(this.defaultArguments) { arguments, entry ->
+        def fileArguments = [makeSingleHyphenArgument(Argument.MAPPING.stylesheet, stylesheetFile)]
```
Comment (eerohele) [inline src/main/groovy/com/github/eerohele/SaxonXsltTask.groovy:834]: What issue does this change fix? Whatever it is, it should be in a different commit.

File: build.gradle [hunk 5]
```diff
@@ -14,7 +14,7 @@ dependencies {
     implementation 'net.sf.saxon:Saxon-HE:9.8.0-14'
     implementation 'xml-resolver:xml-resolver:1.2'
 
-    testCompile('org.spockframework:spock-core:1.3-groovy-2.5') {
+    testImplementation('org.spockframework:spock-core:1.3-groovy-2.5') {
```
Comment (eerohele) [inline build.gradle:17]: https://github.com/eerohele/saxon-gradle/pull/32/commits/0cbe4379b27af1aebc35cc8e93965a07f1167d25 proposes `testCompileOnly` instead of `testImplementation`. Do you know which is correct?

**Review by eerohele — CHANGES_REQUESTED** [review]

File: src/main/groovy/com/github/eerohele/SaxonXsltTask.groovy (hunk 1 above)
Comment (dgeissl) [reply src/main/groovy/com/github/eerohele/SaxonXsltTask.groovy:36]: the problem with having these properties inside the Map is that the map itself contains a bunch of xlt-inputs that influence the outputs and gradle needs to know about them in order to perform a reasonable up-to-date check. Therefor the map itself has to be promoted as `@Input`. But when doing this marking individual keys as for #29 will not be possible (or at least I didn't see a way to do it).

File: src/main/groovy/com/github/eerohele/SaxonXsltTask.groovy (hunk 2 above)
Comment (dgeissl) [reply src/main/groovy/com/github/eerohele/SaxonXsltTask.groovy:178]: I see this makes it difficult to review, sorry about that.

File: src/main/groovy/com/github/eerohele/SaxonXsltTask.groovy (hunk 4 above)
Comment (dgeissl) [reply src/main/groovy/com/github/eerohele/SaxonXsltTask.groovy:834]: as of the gradle 7 deprecation warnings many inputs have to be declared and maps containing file names will always use the absolute path. I sure had #29 in mind when extracting the files to individual properties and therefor the change was necessary.

File: build.gradle (hunk 5 above)
Comment (dgeissl) [reply build.gradle:17]: `testCompileOnly` is something that will not be on the test runtime classpath,
I'm not sure how the spok framework works but my guess is that the tests will complain about missing classes if the dependency is not `testImplementation`. This is also suggested in the spokframework examples https://github.com/spockframework/spock-example/blob/master/build.gradle#L19-L21

**Comment by dgeissl** [comment]
As you already mentioned this pull request is a mix of fixes for #30 and #29.
I guess that #32 is a very simple fix for gradle 7 compatibility and this PR is more about the PathSensivity for inputFiles.

However #32 may have issues when it comes to files. If you declare file properties with `@Input` then only the file path will be the input, but not the content of the file. It looks like `stylesheet` and `inputFiles` are already exposed via a separate getter, so only the `catalog` may be affected - that's not a issue i guess?

File: build.gradle (hunk 5 above)
Comment (jhyry-gcpud) [reply build.gradle:17]: I think dgeissl is correct.
setting it as testCompileOnly was incorrect.

**Comment by eerohele** [comment]
@dgeissl I'm interested in the `@PathSensitivity` changes, so if you'd like to have them, please either rebase on top of the latest `master` and then force-push a new commit that only adds those bits or make a new PR. Ideally, I'd love to have tests for the changes as well.

I cannot see any `@Input`-related deprecation warnings on Gradle 7.2, so I'm not sure whether adding those attributes is necessary after all.

**Comment by dgeissl** [comment]
As mentioned in #29 I currently have no time for working on the PR anymore, that's why I'm closing it for now.
